src/ocean/core/SmartEnum.d

```diff
@@ -786,6 +763,65 @@ unittest
     test!("!in")("NON_EXISTENT".dup, n);
 }
 
+
+/*******************************************************************************
+
+    CTFE function to mixin a SmartEnum class with the codes automatically
+    generated, starting at 0.
+
+    Params:
+        name = name of the class to be created
+        basetype = base type of enum
+        members = variadic list of descriptions of the enum values
+
+*******************************************************************************/
+
+private istring autoSmartEnumImpl ( istring name, istring basetype,
+    istring[] members ... )
+{
+    // Create two strings, one to declare the members, and one to populate
+    // the map.
+    // Also determine the minimum and maximum length of each member name
+
+    istring declstr = null;
+    istring mapstr = null;
+
+    long maxlen = members[0].length;
+    long minlen = members[0].length;
+
+    foreach (i, s ; members )
+    {
+        if (s.length > maxlen)
+        {
+            maxlen = s.length;
+        }
+        if (s.length < minlen )
+        {
+            minlen = s.length;
+        }
+
+        // declstr will be of the form "A=0,B=1,C=2"
+
+        if ( i > 0 ) declstr ~= ",";
+
+        declstr ~= s ~"=" ~ CTFE.toString(i);
+
+        // mapstr will be of the form  `map["A"]=A;map["B"]=B;`
+
+        mapstr ~= `map["` ~ s ~`"]=` ~ s ~ `;`;
+    }
+
+    return  "class " ~ name ~ " : ISmartEnum {" ~
+        " alias " ~ basetype ~ " BaseType; enum : BaseType {"
+        ~ declstr ~ "} enum length = " ~ CTFE.toString(members.length) ~
```

Comment:
you could be more consistent with the position of `~` operator (either on the very begin of newline or in the very end). I personally prefer the operator in the very begin because it is a focused area. I even have something like this in my code:
```d
auto result = first_long_name_variable
    + second_long_name variable;
```

However, whatever you prefer, but consistent.